File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cpp
class Solution {
public:
    
    ListNode * reverseList(ListNode*& head){
        if(!head || !head->next) return head;
        ListNode * prevptr = NULL;
        ListNode * current = head;
        ListNode * nextptr = NULL;
        while(current){
            nextptr = current->next;
            current->next = prevptr;
            prevptr = current;
            current = nextptr;
        }
        return prevptr;
    }
    
    ListNode* reverseKGroup(ListNode* head, int k) {
        if(!head || !head->next) return head;
        int len = 0;
        ListNode * pointHead = head;
        ListNode * pointer = head;
        ListNode * follow = NULL;
        while(pointer && len < k){
            follow = pointer;
            pointer = pointer->next;
            len++;
        }
        if(len < k) return head;
        ListNode * nextHead = pointer;
        follow->next = NULL;
        pointHead = reverseList(pointHead);
        head->next = reverseKGroup(nextHead, k);
        return pointHead;
    }
};

/*

len = 6
     
      -------  
     |       |
0    1  2    3 -> 4 -> 5
p    c  n         
|    |  | 
|    ----
|       |
 --------   
    
0 -> 2 -> 1 -> 3 -> 4 -> 5    
p         c         n     
    
1 -> 4
3 -> 2
0 -> 3    
    
c->next = n->next
n->next = p->next
p->next = n
n = c->next

*/

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.java
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        if(head == null) return head;
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode ptr = dummy;
        int cnt = 0;
        while(ptr != null){
            cnt++;
            ptr = ptr.next;
        }
        ListNode prevptr = dummy;
        ListNode current = dummy;
        ListNode nextptr = dummy;
        while(nextptr != null){
            current = prevptr.next;
            nextptr = current.next;
            int toLoop = (cnt > k ? k : 1);
            for(int i=1;i<toLoop;i++){
                current.next = nextptr.next;
                nextptr.next = prevptr.next;
                prevptr.next = nextptr;
                nextptr = current.next;
            }
            prevptr = current;
            cnt -= k;
        }
        return dummy.next;
    }
}

// 0 -> 3 -> 2 -> 1 -> 4 -> 5 -> 6
// p              c    n    
    
//  c->next = n->next;   
//  n->next = p->next;
//  p->next = n;
//  n = c->next;